bundles/nl.esi.comma.testspecification/src/nl/esi/comma/testspecification/abstspec/generator/RefKVPGenerator.xtend

```diff
@@ -38,24 +34,19 @@ import nl.esi.comma.assertthat.assertThat.ScriptParameterWithValue
 import nl.esi.comma.assertthat.assertThat.ScriptParameterNamedArg
 import nl.esi.comma.assertthat.assertThat.ScriptParameterPositional
 import nl.esi.comma.assertthat.assertThat.ScriptParameterPositionalFile
+import nl.esi.comma.expressions.expression.ExpressionBracket
+import nl.esi.comma.testspecification.testspecification.RunStep
 
 class RefKVPGenerator {
     def generateRefKVP(AbstractTestDefinition atd) {
         var txt = 
         '''
-            «FOR test : atd.testSeq»
-            «generateMatlabCallFragment(getScriptCalls(test.step))»
-
-            «generateAssertionsFragment(getAssertionItems(test.step))»
-            «ENDFOR»
-        '''
-        return txt
-    }
-
-    def generateMatlabCallFragment(ArrayList<GenericScriptBlock> blocks) {
-        return '''
         matlab_calls=[
-            «FOR mlcal : blocks SEPARATOR ',' »
+        «FOR testseq : atd.testSeq»
+            «FOR step : testseq.step.filter(AssertionStep) »
+            «FOR asrtce : step.asserts » 
+            «FOR ce : asrtce.ce»
+                «FOR mlcal : ce.constr.filter(GenericScriptBlock) SEPARATOR ',' »
```

Comment:
Please explore the lambda expressions in Xtend a bit more, and don't create a template FOR loop for every iteration.

```suggestion
                «FOR mlcal : step.asserts.flatMap[ce].flatMap[constr].filter(GenericScriptBlock) SEPARATOR ',' »
```